Support Lora Block Weight (LBW) to svd_merge_lora.py

terracottahaniwa opened the pull request:
I had added support that lora block weight to svd_merge_lora.py.
- --sd2 and --lbws options are added.
- --sd2 option is used to parse key names from diffusers format to compvis format.
- --lbws is used to pass the block weights as follows;
```
python svd_merge_lora.py --save_precision bf16 --precision float \
--save_to output.safetensors --models input1.safetensors input2.safetensors \
--ratios 1 1 --lbws "[1,1,1,1,1,1,1,1,1,1,1,1]" "[1,1,1,1,1,1,1,1,1,1,1,1]" --new_rank 1
```

JujoHotaru commented:
XL対応のLoRAやデータモデル開発を行っている十条と申します。  
実はこのPR、SuperMergerのXL LoRAの階層マージがうまく動かず困っていた私がDiscordで相談し、実装していただいたものでした。対応検討いただけるとのことで期待しております。

terracottahaniwa commented:
Thanks for the reply.
I see, I will going to try to solve the license incompatibility issue.

terracottahaniwa commented:
I rewrote the code using the `get_block_index()` function. During the code review, I found that the code does not depend on whether the model is SD2 or Not. Therefore, I removed the `--sd2` option again.

terracottahaniwa commented:
I also noticed that Hakomikan's layer selection criteria is to have attention blocks.
```py
lora = safetensors.safe_open(r"sdxl.safetensors", framework="pt")
s = []
for key in lora.keys():
    if "lora_down" not in key:
        continue
    if "attn" in key:
        m = re.search(r"_(\d+)_(\d)_", key)
        if m:
            s.append(key[0:m.end()])
print(sorted(list(set(s))))
"""
[
    'lora_unet_input_blocks_4_1_', 
    'lora_unet_input_blocks_5_1_', 
    'lora_unet_input_blocks_7_1_', 
    'lora_unet_input_blocks_8_1_', 
    'lora_unet_output_blocks_0_1_', 
    'lora_unet_output_blocks_1_1_', 
    'lora_unet_output_blocks_2_1_', 
    'lora_unet_output_blocks_3_1_', 
    'lora_unet_output_blocks_4_1_', 
    'lora_unet_output_blocks_5_1_'
]
"""
```
So, maybe, BASE has 'encoder', MID has 'middle', and other target layers have 'attn'. Thus, if the key contains strings like 'lora_unet_input_blocks_4_1_', Is it okay to consider this key as the target?

alternative implementation
https://github.com/kohya-ss/sd-scripts/compare/main...terracottahaniwa:sd-scripts:dev

JujoHotaru commented:
本件ご対応進めていただいておりますところですが、LoRA同士のマージのほか、XLモデルへのLoRAマージ時（sdxl_merge_lora.py）でもlbws指定ができるとデータモデル開発において大変はかどります。
よろしければそちらもご検討いただければ幸いです。

terracottahaniwa commented:
Thank you for correcting the function! It seems to work well.
I rewrote the code again using `get_lbw_block_index()`.

About the `sdxl_merge_lora.py`, there's another pull request (#1580) currently in progress.
So maybe we should wait a bit for that PR to resolve as well.

JujoHotaru commented:
ご確認いただいありがとうございます。
書き忘れておりましたが対象はSDXLのLoRAで、実は上記コメントの「svd_merge_loraに与える引数」のところに記載しているように、すでに不足分は0で埋めて20個分のlbwとしたものを引数に与えておりました。（LoRA BlockWeightの動作に合わせるため）
SDXLなので20個でよいかと思うのですが、26個指定したほうがよいでしょうか？ひとまず試してみます。

JujoHotaru commented:
lbw26個指定のLoRAを作ってみましたが、出力結果はlbw20個指定で出したものと同じでした。

JujoHotaru commented:
かしこまりました。早速ログを取ってみました。（添付）

[log.txt](https://github.com/user-attachments/files/17045105/log.txt)

「AAAA:2:0,0,1,1,1,1,1,1,1,0,0,0」を入力とした場合のデータになります。試しに強度は2.0にしてあります。

JujoHotaru commented:
なお、もし現物データがあったほうがよいようでしたらご提供可能です。

terracottahaniwa commented:
Oops, maybe my mistake.
The weight might multiply only up block.

https://github.com/hako-mikan/sd-webui-lora-block-weight/blob/4d94d247a6e6d77c11c2d536a5138a54f4d9e500/scripts/lora_block_weight.py#L1123

terracottahaniwa commented:
I have researched this in the past when making my tools, but it seems that time has affected my memory...

https://github.com/terracottahaniwa/apply-lora-block-weight/blob/main/apply_lora_block_weight.py#L80-L81

terracottahaniwa commented:
I apologize. I have incorrectly recalled the details of the research conducted during made my tool...
In that tool's implementation, it had multiplied to only the up weights.

https://github.com/terracottahaniwa/apply-lora-block-weight/blob/main/apply_lora_block_weight.py#L80-L81

JujoHotaru commented:
早速最新コミットで試してみました。
結果、4つのLoRAを指定していた場合とほぼ同等の出力結果が得られる合成LoRAになりました！無事に修正できているようです。
調査・ご対応ありがとうございました。効率上昇で開発ペースを上げられそうです。

JujoHotaru commented:
ちなみに、sdxl_merge_loraでモデルにLoRAをマージする場合についても同様の問題はございましたでしょうか？

JujoHotaru commented:
承知しました。引き続き活用させていただきます。

JujoHotaru commented:
追伸ですが、今回修正版でマージしたLoRAは、私の最初の報告の最後の手順（あらかじめSuperMergerで階層だけ制限しておいたLoRAを強度だけ指定でマージ）で作ったLoRAとハッシュが完全一致しておりました。
理論的には同じ計算手順ですので順当なのだと思いますが、一応ご報告させていただきます。

terracottahaniwa commented:
Simply, I need to relearning! 😂
![image](https://github.com/user-attachments/assets/1db61023-a53d-4be2-8152-3b764e4c9bf8)
